plugins/module_utils/options.py

```diff
@@ -0,0 +1,106 @@
+# coding: utf-8 -*-
+# GNU General Public License v3.0+
+# (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import (absolute_import, division, print_function)
+
+__metaclass__ = type
+
+from dataclasses import dataclass, field
+from typing import List, Optional
+
+from .constants import Kind
+from .exceptions import ValidationException
+
+
+@dataclass
+class ApprovalOptions:
+    kind: Kind
+    name: str
+    approved: bool
+    approved_by: Optional[str] = None
+    labels: Optional[dict] = None
+
+    def __post_init__(self):
+        if self.kind not in [Kind.CSR, Kind.ENROLLMENT]:
+            raise ValidationException(f"Kind {self.kind.value} does not support approval")
+        if not self.name:
+            raise ValidationException("Name must be specified")
+        if self.approved is None:
+            raise ValidationException("Approved must be specified")
+
+    def to_request_params(self):
+        params = dict(
+            approved=self.approved,
+        )
+        if self.approved_by:
+            params['approvedBy'] = self.approved_by
+        if self.labels:
+            params['labels'] = self.labels
+        return params
+
+
+@dataclass
+class GetOptions:
+    kind: Kind
+    name: Optional[str] = None
+    label_selector: Optional[str] = None
+    field_selector: Optional[str] = None
+    owner: Optional[str] = None
+    fleet_name: Optional[str] = None
+    rendered: Optional[bool] = None
+    summary: Optional[bool] = None
+    summary_only: Optional[bool] = None
+    status_filter: List[str] = field(default_factory=list)
+    limit: Optional[int] = None
+    continue_token: Optional[str] = None
+
+    def __post_init__(self):
+        if not self.kind:
+            raise ValidationException("Kind must be specified")
+        if self.owner and self.kind not in [Kind.DEVICE, Kind.FLEET]:
+            raise ValidationException("Owner field is only valid for Device and Fleet kinds")
+        if self.rendered and self.kind is not Kind.DEVICE:
+            raise ValidationException("Rendered field is only valid for Device kind")
+        if self.fleet_name and self.kind is not Kind.TEMPLATE_VERSION:
+            raise ValidationException("Fleet name field is only valid for TemplateVersion kind")
+        if self.summary_only:
+            if self.kind is not Kind.DEVICE:
+                raise ValidationException("Summary Only field is only valid for Device kind")
+            if self.name:
+                raise ValidationException("Summary Only field is not valid when fetching one Device")
+        if self.summary:
```

Comment:
I don't love how this looks w/ so many fields and validations.  This pr is getting quite large but I will probably try to come back and refactor, possibly consolidating some of the logic at the field level or offload this to the ansible argument spec validator https://docs.ansible.com/ansible/latest/reference_appendices/module_utils.html#argumentspecvalidator